### DISCUSSION TITLE
Irregular shaped cellular automata space

### DISCUSSION THREAD

**Discussion by clarkpetri** [mesa/mesa#1188] [Q&A] [answered]
Are there recommendations for shaping a grid beyond m x n? I'm specifically working on an evacuation model and want to funnel agents through a narrow space (door). Imagine an m x n space connected to another m x n space via a 1x2 space.

From my (admittedly meager) understanding of Mesa I suspect I can hard code in agents that don't move within a larger and fixed 2m x 2n space, using a SingleGrid, effectively forcing the desired effect. **I'm wondering if there's a more elegant solution**.

Also, any suggestions for encouraging movement toward the "exit" would be appreciated. I understand what would be needed for a deterministic and fixed approach, mathematically ensuring agents only move down from the upper space for example, but again -- open to suggestions and learning.

Thank you! My academic curriculum is generally all about NetLogo but as a Python fan I'm trying to blaze a trail with Mesa.

**Comment by clarkpetri** [comment, marked as answer]
I created a work-around that serves my purpose. By instigating "wall agents" which don't interact and aren't added to the schedule I can explicitly place them within the m x n to create the shapes I want. While this will only work within a single grid structure and is tedious to shape larger areas, it does allow structuring cellular automata within Mesa outside of strict m x n.

**Reply by rht** [reply]
You could also make arbitrary topology with the `NetworkGrid`. Not sure if this is more performant than fixture "agents".
